src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java

```diff
@@ -2760,49 +2757,93 @@ static int getSurplusQueuedTaskCount() {
+        if (now) {
+            releaseWaiters();
+            for (;;) {
+                if (((e = runState) & CLEANED) == 0L) {
+                    boolean clean = cleanQueues();
+                    if (((e = runState) & CLEANED) == 0L && clean)
+                        e = getAndBitwiseOrRunState(CLEANED) | CLEANED;
+                }
+                if ((e & TERMINATED) != 0L)
+                    break;
+                if (ctl != 0L) // else loop if didn't finish cleaning
+                    break;
+                if ((e & CLEANED) != 0L) {
+                    e |= TERMINATED;
+                    if ((getAndBitwiseOrRunState(TERMINATED) & TERMINATED) == 0L) {
+                        CountDownLatch done; SharedThreadContainer ctr;
+                        if ((done = termination) != null)
+                            done.countDown();
+                        if ((ctr = container) != null)
+                            ctr.close();
                     }
+                    break;
                 }
             }
-            if (((e = runState) & TERMINATED) == 0L && ctl == 0L) {
-                e |= TERMINATED;
-                if ((getAndBitwiseOrRunState(TERMINATED) & TERMINATED) == 0L) {
-                    CountDownLatch done; SharedThreadContainer ctr;
-                    if ((done = termination) != null)
-                        done.countDown();
-                    if ((ctr = container) != null)
-                        ctr.close();
+        }
+        return e;
+    }
+
+    /**
+     * Scans queues in a psuedorandom order based on thread id,
+     * cancelling tasks until empty, or returning early upon
+     * interference or still-active external queues, in which case
+     * other calls will finish cancellation.
+     *
+     * @return true if all queues empty
+     */
+    private boolean cleanQueues() {
+        int r = (int)Thread.currentThread().threadId();
+        r ^= r << 13; r ^= r >>> 17; r ^= r << 5; // xorshift
+        int step = (r >>> 16) | 1;                // randomize traversals
+        WorkQueue[] qs = queues;
+        int n = (qs == null) ? 0 : qs.length;
+        for (int l = n; l > 0; --l, r += step) {
+            WorkQueue q; ForkJoinTask<?>[] a; int cap;
+            if ((q = qs[r & (n - 1)]) != null &&
+                (a = q.array) != null && (cap = a.length) > 0) {
+                for (;;) {
+                    ForkJoinTask<?> t; int b; long k;
+                    t = (ForkJoinTask<?>)U.getReferenceAcquire(
+                        a, k = slotOffset((cap - 1) & (b = q.base)));
+                    if (q.base == b && t != null &&
+                        U.compareAndSetReference(a, k, t, null)) {
+                        q.updateBase(b + 1);
+                        try {
+                            t.cancel(false);
+                        } catch (Throwable ignore) {
+                        }
+                    }
+                    else if ((q.phase & (IDLE|1)) == 0 ||
```

Comment:
@DougLea What's the intent of the `|1` part here? 🤔

Reply:
clarified by:
                    else if ((q.phase & (IDLE|1)) == 0 || // externally locked